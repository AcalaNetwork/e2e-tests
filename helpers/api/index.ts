import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
