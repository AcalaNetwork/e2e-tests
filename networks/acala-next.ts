import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
