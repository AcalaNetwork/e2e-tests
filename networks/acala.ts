import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
