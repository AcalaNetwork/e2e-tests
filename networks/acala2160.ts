import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
