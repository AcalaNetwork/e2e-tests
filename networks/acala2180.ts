import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
