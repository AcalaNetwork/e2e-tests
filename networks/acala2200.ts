import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
