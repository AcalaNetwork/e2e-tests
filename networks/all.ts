import { Config } from './types'

import acala2160Config from './acala2160'
import acalaConfig from './acala'
import hydraDXConfig from './hydraDX'
import moonbeamConfig from './moonbeam'
import polkadot9381Config from './polkadot9381'
import polkadotConfig from './polkadot'
import statemintConfig from './statemint'

const all = {
  polkadot: polkadotConfig,
  polkadot9381: polkadot9381Config,
  statemint: statemintConfig,
  acala: acalaConfig,
  acala2160: acala2160Config,
  moonbeam: moonbeamConfig,
  hydraDX: hydraDXConfig,
} satisfies Record<string, Config>

export default all
