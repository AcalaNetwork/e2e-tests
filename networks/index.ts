import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
