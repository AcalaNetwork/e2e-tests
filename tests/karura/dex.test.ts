import { afterAll, beforeEach, describe, it } from 'vitest'

import { expectJson, matchEvents, sendTransaction, testingPairs } from '../helper'
import { queryTokenBalance } from '../api/query'
import { stableAssetSwap, swapWithExactSupply, swapWithExactTarget } from '../api/extrinsics'
import networks from '../networks'

describe('Karura <-> Kusama', async () => {
  const karura = await networks.karura()

  const { alice } = testingPairs()

  afterAll(async () => {
    await karura.teardown()
  })

  beforeEach(async () => {
    await karura.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: 'KSM' }], { free: 100 * 1e12 }],
          [[alice.address, { Token: 'LKSM' }], { free: 1000 * 1e12 }],
          [[alice.address, { Token: 'KUSD' }], { free: 0 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    })
  })

  it('supply swap works', async () => {
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 0,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 100000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    const tx = await sendTransaction(
      swapWithExactSupply(karura.api, [{ Token: 'KSM' }, { Token: 'KUSD' }], '1000000000000', '0').signAsync(alice)
    )
    await karura.chain.newBlock()
    await matchEvents(tx.events, {
      method: 'Swap',
      section: 'dex',
    })
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 51959600247621,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 99000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
  })

  it('target swap works', async () => {
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 0,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 100000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    const tx = await sendTransaction(
      swapWithExactTarget(
        karura.api,
        [{ Token: 'KSM' }, { Token: 'KUSD' }],
        '1000000000000',
        '10000000000000'
      ).signAsync(alice, { nonce: 0 })
    )
    await karura.chain.newBlock()
    await matchEvents(tx.events, {
      method: 'Swap',
      section: 'dex',
    })
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 1000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 99980750049439,
        "frozen": 0,
        "reserved": 0,
      }
    `)
  })

  it('stable swap works', async () => {
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 0,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 100000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    const tx0 = await sendTransaction(
      swapWithExactTarget(
        karura.api,
        [{ Token: 'KSM' }, { Token: 'KUSD' }],
        '1000000000000',
        '10000000000000'
      ).signAsync(alice, { nonce: 0 })
    )
    await karura.chain.newBlock()
    await matchEvents(tx0.events, {
      method: 'Swap',
      section: 'dex',
    })
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 1000000000000,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { Token: 'KSM' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 99980749889928,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    const tx1 = await sendTransaction(
      stableAssetSwap(karura.api, '1', '0', '2', '1000000000000', '0', '3').signAsync(alice, { nonce: 1 })
    )
    await karura.chain.newBlock()
    await matchEvents(tx1.events, { method: 'TokenSwapped', section: 'stableAsset' })
    expectJson(await queryTokenBalance(karura.api, { Token: 'KUSD' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 0,
        "frozen": 0,
        "reserved": 0,
      }
    `)
    expectJson(await queryTokenBalance(karura.api, { ForeignAsset: '7' }, alice.address)).toMatchInlineSnapshot(`
      {
        "free": 722149,
        "frozen": 0,
        "reserved": 0,
      }
    `)
  })
})
