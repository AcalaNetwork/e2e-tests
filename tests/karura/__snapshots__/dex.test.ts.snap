// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Karura <-> Kusama > stable swap works > events 1`] = `
[
  {
    "data": {
      "liquidityChanges": [
        "(redacted)",
        "(redacted)",
      ],
      "path": [
        {
          "Token": "KSM",
        },
        {
          "Token": "KUSD",
        },
      ],
      "trader": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "Swap",
    "section": "dex",
  },
]
`;

exports[`Karura <-> Kusama > stable swap works > events 2`] = `
[
  {
    "data": {
      "a": "(redacted)",
      "balances": [
        "(redacted)",
        "(redacted)",
        "(redacted)",
      ],
      "inputAmount": "(redacted)",
      "inputAsset": {
        "Token": "KUSD",
      },
      "minOutputAmount": "(redacted)",
      "outputAmount": "(redacted)",
      "outputAsset": {
        "ForeignAsset": "(redacted)",
      },
      "poolId": "(redacted)",
      "swapper": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
      "totalSupply": "(redacted)",
    },
    "method": "TokenSwapped",
    "section": "stableAsset",
  },
]
`;

exports[`Karura <-> Kusama > supply swap works > events 1`] = `
[
  {
    "data": {
      "liquidityChanges": [
        "(redacted)",
        "(redacted)",
      ],
      "path": [
        {
          "Token": "KSM",
        },
        {
          "Token": "KUSD",
        },
      ],
      "trader": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "Swap",
    "section": "dex",
  },
]
`;

exports[`Karura <-> Kusama > target swap works > events 1`] = `
[
  {
    "data": {
      "liquidityChanges": [
        "(redacted)",
        "(redacted)",
      ],
      "path": [
        {
          "Token": "KSM",
        },
        {
          "Token": "KUSD",
        },
      ],
      "trader": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "Swap",
    "section": "dex",
  },
]
`;
